### PR TITLE
Adopt announcement settings

### PR DESCRIPTION
**Pull request by allenwq**
Depends on #409

### CODE REVIEW

File: app/controllers/course/announcements_controller.rb [hunk 1]
```diff
@@ -51,4 +52,18 @@ def destroy #:nodoc:
   def announcement_params #:nodoc:
     params.require(:announcement).permit(:title, :content, :sticky, :valid_from, :valid_to)
   end
+
+  # Ensure that the component is enabled and load its settings.
+  #
+  # @raise [ActionController::RoutingError] When the component is disabled.
+  def check_component_and_load_settings
+    announcement_component = current_component_host[:course_announcements_component]
+    raise ActionController::RoutingError.
```
Comment (allenwq) [inline app/controllers/course/announcements_controller.rb]: @lowjoel Am I raising the correct error ? basically I want rails to show a "page does not exists" page when component was disabled.

**Comment by lowjoel** [comment]
Rebase please.

File: lib/autoload/coursemology.rb [hunk 2]
```diff
@@ -0,0 +1,7 @@
+module Coursemology
+  class ComponentNotFoundError < StandardError
```
Comment (lowjoel) [inline lib/autoload/coursemology.rb]: No need to nest within the Coursemology module.

File: app/controllers/course/announcements_controller.rb [hunk 3]
```diff
@@ -51,4 +52,18 @@ def destroy #:nodoc:
   def announcement_params #:nodoc:
     params.require(:announcement).permit(:title, :content, :sticky, :valid_from, :valid_to)
   end
+
+  # Ensure that the component is enabled and load its settings.
+  #
+  # @raise [ActionController::RoutingError] When the component is disabled.
```
Comment (lowjoel) [inline app/controllers/course/announcements_controller.rb]: Wrong exception.

File: lib/autoload/coursemology.rb (hunk 2 above)
Comment (allenwq) [inline lib/autoload/coursemology.rb]: Seems like it's a convention to scope exceptions under project name ? Like 'Cancan::AccessDenied'

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: That's different, it's a gem.

Comment (allenwq) [inline lib/autoload/coursemology.rb]: I also saw some other rails projects doing this, like: https://github.com/discourse/discourse/blob/master/lib/discourse.rb#L31

How about put them under exceptions ? (http://stackoverflow.com/questions/5200842/where-to-define-custom-error-types-in-ruby-and-or-rails)

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: I personally find it superfluous.

Comment (allenwq) [inline lib/autoload/coursemology.rb]: Then what's your suggestion ? I think all custom exceptions should be scoped under a file or a folder

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: I would leave it in global scope.

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: Honestly if you want to put it somewhere, putting it in a top-level application module is no different from putting it in the global scope. Find some logical place to put it if you want to have some sort of compartmentalisation.

Comment (allenwq) [inline lib/autoload/coursemology.rb]: Yah, basically what I want is compartmentalisation.

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: Then putting it in the top level module doesn't serve any purpose. Place it in the component host or something.

Comment (allenwq) [inline lib/autoload/coursemology.rb]: How about `Componentize::ComponentNotFound`, that is `lib/autoload/componentize.rb`

Comment (lowjoel) [inline lib/autoload/coursemology.rb]: nope, not all componentized modules would raise this error.

File: app/controllers/course/announcements_controller.rb [hunk 4]
```diff
@@ -51,4 +52,18 @@ def destroy #:nodoc:
   def announcement_params #:nodoc:
     params.require(:announcement).permit(:title, :content, :sticky, :valid_from, :valid_to)
   end
+
+  # Ensure that the component is enabled and load its settings.
+  #
+  # @raise [Coursemology::ComponentNotFoundError] When the component is disabled.
+  def check_component_and_load_settings
```
Comment (lowjoel) [inline app/controllers/course/announcements_controller.rb]: I would separate this into a separate class for more complex components, and all controllers in that class would inherit from that controller.

Comment (lowjoel) [inline app/controllers/course/announcements_controller.rb]: Maybe there should be a `#component` method on this group of controllers?

File: lib/autoload/component_not_found_error.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class ComponentNotFoundError < StandardError
+  def message
```
Comment (lowjoel) [inline lib/autoload/component_not_found_error.rb]: this should be set in an initializer and passed to super. remember to freeze the string.

Comment (allenwq) [inline lib/autoload/component_not_found_error.rb]: Thanks, updated !

**Pull request merged**